Convert routes to a node tree to improve route matching perfomance

marcelthole opened the pull request:
Hey,
we added this router in one of our projects and noticed a perfomance problem with this change.

With this PR i introduced a TreeNode array for the Map that will put all possible routes per route segment. 
This will reduce the amount of Routes that have to be checked to the routes that match at least in the segment names. The specific matching with HTTP Verb, and validating the dynamic properties in the URL will be applied as before for the matched routes. 

I also created this phpbench File to compare the Results:

**MatchBench.php**
```php
<?php

namespace Aura\Router\Benchmark;

use Aura\Router\RouterContainer;
use GuzzleHttp\Psr7\ServerRequest;
use Psr\Http\Message\ServerRequestInterface;

/**
 * @BeforeMethods("setUp")
 */
class MatchBench
{
    /** @var RouterContainer $container */
    private $container;

    public function setUp()
    {
        $this->container = new RouterContainer();
        $map = $this->container->getMap();

        foreach ($this->routesProvider() as $key => $route) {
            $map->get($key, $route, static function () use ($key) { return $key; });
        }
    }

    /**
     * @Iterations(3)
     */
    public function benchMatch()
    {
        $matcher = $this->container->getMatcher();
        foreach ($this->routesProvider() as $route) {
            $result = $matcher->match($this->stringToRequest($route));
            if ($result === false) {
                throw new \RuntimeException(sprintf('Expected route "%s" to be an match', $route));
            }
        }
    }

    private function routesProvider()
    {
        $segments = ['one', 'two', 'three', 'four', 'five', 'six'];
        $routesPerSegment = 100;

        $routeSegment = '';
        foreach ($segments as $index => $segment) {
            $routeSegment .= '/' . $segment;
            for ($i = 1; $i <= $routesPerSegment; $i++) {
                yield $index . '-' . $i => $routeSegment . $i;
            }
        }
    }

    /**
     * @param string $url
     * @return ServerRequestInterface
     */
    private function stringToRequest($url)
    {
        return new ServerRequest('GET', $url, [], null, '1.1', []);
    }
}
```

**Result**:
```
/var/www # vendor/bin/phpbench run tests/Benchmark/ --ref=current
PHPBench (1.4.1) running benchmarks... #standwithukraine
with configuration file: /var/www/phpbench.json
with PHP version 8.4.1, xdebug ✔, opcache ❌
comparing [actual vs. current]

\Aura\Router\Benchmark\MatchBench

    benchMatch..............................I2 - [Mo828.080ms vs. Mo1.734s] -52.25% (±0.95%)

Subjects: 1, Assertions: 0, Failures: 0, Errors: 0
```

i also played with some caching for the generated tree route and got this result. But this is not part of this PR. i would create a followup PR for this if this PR is fine for you. 

```
/var/www # vendor/bin/phpbench run tests/Benchmark/ --ref=current
PHPBench (1.4.1) running benchmarks... #standwithukraine
with configuration file: /var/www/phpbench.json
with PHP version 8.4.1, xdebug ✔, opcache ❌
comparing [actual vs. current]

\Aura\Router\Benchmark\MatchBench

    benchMatch..............................I2 - [Mo17.109ms vs. Mo1.734s] -99.01% (±1.49%)

Subjects: 1, Assertions: 0, Failures: 0, Errors: 0
```

I added this Benchmark here in the PR description because PHPBench does not support PHP 5.5 anymore.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Improved route matching performance by organizing routes into a hierarchical tree structure based on path segments.
  - Added a comprehensive benchmarking script to measure route matching speed and memory usage across various route types and counts.
- **Bug Fixes**
  - Enhanced handling of non-routable routes to ensure they are correctly excluded during matching.
- **Tests**
  - Added benchmark tests to measure performance of route generation and matching.
  - Introduced new tests to verify the correctness of the tree-based route organization and matching logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

marcelthole committed:
Convert routes to a node tree to improve route matching perfomance

marcelthole commented on src/Matcher.php:
```diff
@@ -131,20 +135,18 @@ public function match(ServerRequestInterface $request)
     {
         if (! $proto->isRoutable) {
-            return;
+            return false;
```
See doctype description: `@return mixed False on failure, or a Route on match.`
so this should return false?

marcelthole commented on src/Matcher.php:
```diff
@@ -131,20 +135,18 @@ public function match(ServerRequestInterface $request)
         }
         $route = clone $proto;
-        return $this->applyRules($request, $route, $name, $path);
+        return $this->applyRules($request, $route, $route->name, $path);
```
routename is already available in the route itself

marcelthole commented on src/Matcher.php:
```diff
@@ -261,4 +263,27 @@ public function getMatchedRoute()
+                $node = $node[$segment];
+                continue;
+            }
+            if (isset($node['{}'])) {
```
Group all kind of routes with dynamic properties into a generic group

marcelthole commented on tests/MatcherTest.php:
```diff
@@ -258,7 +258,6 @@ public function testLogger()
         $matcher->match($request);
 
         $expect = [
-            'debug: /bar FAILED Aura\Router\Rule\Path ON foo',
```
This will not work anymore because the /foo route will not be checked anymore.
Is this a problem for you? Do it really need hundreds of debug messages that a specific route didn't matched?

harikt commented:
Hey @marcelthole 

I have noticed this PR, but not sure how the phpbench was checking the current code ( in 3.x branch ) vs the new code in the PR.

@koriym @pmjones what are your thoughts on this ?

harikt commented:
Another option you could do is create a different matcher class and use it for now.

Optionally : We can add support of passing different Matcher to the constructor of the RouteContainer.

marcelthole commented:
To the phpbench topic: 
In this case i created the bench Testfile and stored the result and compared it later with this patch. 
https://phpbench.readthedocs.io/en/latest/guides/regression-testing.html#comparison
For later steps it could be also done via github workflow to create the branch in the target branch and after that run the benchmark again with the changed files. 

And yes for now i could use my own Matcher for that in our codebase. But if this feature could improve the perfomance of the router for larger route configurations it would be a benefit for other people too :)

harikt commented:
I believe this https://github.com/auraphp/Aura.Router/pull/207 will improve the performance. Have you checked this ?

I will wait for others input for this PR for the current time.

marcelthole committed:
Add Unittests for new changes

marcelthole commented:
#207 will only help in the case where i would generate a route if i saw it correctly. That has no impact on the bootstrapping time to match a route. 

But did also a Bench for that change:
```php

/**
 * @BeforeMethods("setUp")
 */
class GeneratorBench
{
    /**
     * @var Generator
     */
    private $generator;

    public function setUp()
    {
        $map = new Map(new Route());
        foreach ($this->routesProvider() as $key => $route) {
            $map->get($key, $route, static function () use ($key) { return $key; });
        }

        $map->get('dummy', '/api/user/{id}/{action}/{controller:[a-zA-Z][a-zA-Z0-9_-]{1,}}{/param1,param2}');
        $this->generator = new Generator($map);
    }


    private function routesProvider()
    {
        $segments = ['one', 'two', 'three', 'four', 'five', 'six'];
        $routesPerSegment = 100;

        $routeSegment = '';
        foreach ($segments as $index => $segment) {
            $routeSegment .= '/' . $segment;
            for ($i = 1; $i <= $routesPerSegment; $i++) {
                yield $index . '-' . $i => $routeSegment . $i;
            }
        }
    }


    /**
     * @Revs(1000)
     * @Iterations (10)
     */
    public function benchMatch()
    {
        $this->generator->generate('dummy', [
            'id' => 1,
            'action' => 'doSomethingAction',
            'controller' => 'My_User-Controller1',
            'param1' => 'value1',
            'param2' => 'value2',
        ]);
    }
}
```

```
\Aura\Router\Benchmark\GeneratorBench

    benchMatch..............................I9 - [Mo8.025μs vs. Mo8.353μs] -3.92% (±3.11%)
```

Yes it will be a little bit faster, but will not help me at all.
But for now i added a own Implementation for that in our codebase and added some unit tests for the changes aswell

koriym commented:
Test the code rabbit. @coderabbitai review

koriym commented:
@coderabbitai review

koriym committed:
Add Aura.Router benchmark script for performance analysis

This script benchmarks route matching performance and memory usage for varying numbers of routes. It includes features like baseline comparison, support for route generation with different types, and formatted output for detailed analysis.

koriym committed:
Revert "Add Aura.Router benchmark script for performance analysis"

This reverts commit 66560e67f27c25ced52703af2f696dc7370a5744.

koriym commented:
Sorry, I pushed the benchmark script here by mistake, but I revert it.

koriym commented:
I have done an AI benchmarking program and analysis. I think each is reasonable.
[Aura.Router Performance Benchmark Analysis for PR#208](https://github.com/koriym/Aura.Router/issues/2)

pmjones commented on tests/Benchmark/MatchBench.php:
```diff
@@ -0,0 +1,70 @@
+     */
+    public function benchMatch()
+    {
+        $this->container->getMap()->treeRoutes = $this->treeNodes;
+        $matcher = $this->container->getMatcher();
+        foreach ($this->routesProvider() as $route) {
+            $result = $matcher->match($this->stringToRequest($route));
+            if ($result === false) {
+                throw new \RuntimeException(sprintf('Expected route "%s" to be an match', $route));
+            }
+        }
```
Unless I'm missing something, I don't think this benchmark is built properly. Because matching is only going to happen once per request, it makes more sense to benchmark the complete setting-up and then the matching. That is: start the timer, start a loop, set up the container & routes, match a route, repeat the loop, then stop the timer when the loop is over. That will give a better sense of the comparative speed. Cf. https://github.com/pmjones/AutoRoute-benchmark/blob/2.x/bin/benchmark.php

marcelthole replied:
Yes you are right, this will bench the best case after a caching logic for the tree node would be implemented. I will update this with a second bench where the node tree will be regenerated for every match call

- [ ] Add a bench for the case where the treenode must be regenerated for every match call

marcelthole commented:
ah i guess i also found a bug here. 
I need to add a test for routings like `/api/export/country{/countryId}.csv` that this will match `/api/export/country.csv` and `/api/export/country/1.csv` (same behavior for `/api/export.{format}` for `/api/export.csv` and `/api/export.jsonl`

- [ ] Add tests for the two examples

harikt commented:
@marcelthole did you noticed @pmjones comment ? If not https://github.com/auraphp/Aura.Router/pull/208#discussion_r2072398254
